### PR TITLE
Add option to specify if the customer choose to receive available products

### LINKED ISSUE #9781: [BOOM-4539] Delayed shipping & out of stock product
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-4539).

- _**Reporter:**_ marion_francois
- _**Created at:**_ Thu, 28 Dec 2017 17:32:54 +0100

I enable in BO the order of out-of-stock product (in product settings) and the delayed shipping option (in order settings)

I add to my cart an out of stock product (with quantity < 0) and an in stock product (with quantity > 0)

In 1.6, in front office I could check a box when ordering to receive the available products first.

Capture du 2017-12-28 16-43-16

The order then was split in to 2 BO orders: 
- One with products in stock 
- The other with the out of stock products

This no longer works in 1.7 because the FO checkbox is no longer displayed.

The idea is to restore the checkbox in FO and therefore the order split.

The checkbox should therefore only appear if: 
- delayed shipping is enabled 
- ordering products out of stock is allowed 
- the cart contains at least one product with a stock <= 0

### PR DESCRIPTION
**Pull request by thalassa-web**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | In PrestaShop 1.6 the customer was able to choose to recieve available products first. The option is still available in cart but the controller do not use it anymore.
| Type?             | bug fix
| Category?         |  CO
| BC breaks?        | no
| Deprecations?     |  no
| Fixed ticket?     | Fixes #9781
| How to test?      | Add an available product and an unavailable product (but orderable) in the cart. Enter in the checkout tunnel. In the shipping step, check the checkbox "Send available products first". The order should be split in two.
| Possible impacts? | The option is inactive by default, impacts are very limited.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/22941)

### CODE REVIEW

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (matks) [inline themes/classic/templates/checkout/_partials/steps/shipping.tpl]:
```suggestion
                <input class="js-gift-checkbox" id="allow_separated_package" name="allow_separated_package"
```

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (matks) [inline themes/classic/templates/checkout/_partials/steps/shipping.tpl]:
```suggestion
                <label for="allow_separated_package">{l s='Send available products first' d='Shop.Theme.Checkout'}</label>
```

File: classes/checkout/CheckoutDeliveryStep.php
Comment (matks) [inline classes/checkout/CheckoutDeliveryStep.php]:
```suggestion
                isset($requestParams['allow_separated_package']) ? $requestParams['allow_separated_package'] : false
```
Little typo 😉

File: classes/checkout/CheckoutDeliveryStep.php
Comment (matks) [inline classes/checkout/CheckoutDeliveryStep.php]: Hi, I think it could be useful to add phpDoc 😊  ?

I don't know if we can type arguments and return type 🤔  wdyt @Progi1984 ? should we introduce strong type hint even if rest of the class does not use it ?

File: classes/checkout/CheckoutSession.php
Comment (matks) [inline classes/checkout/CheckoutSession.php:202]:
```suggestion
        $this->context->cart->allow_seperated_package = (bool) $separatePackagesAllowed;
```
If we do not force the type to bool, maybe we should cast it here to make sure we don't mess the cart 👀

Comment (matks) [reply classes/checkout/CheckoutSession.php:202]: Also typo again 😅 
```suggestion
        $this->context->cart->allow_separated_package = $separatePackagesAllowed;
```

**Review by matks — COMMENTED** [review]
Hi @thalassa-web thank you for the PR it is a lot better now 🎉 

I added a few suggestions to make the code even better.

So about the behavior: this PR makes the option
- visible in checkout (customer can decide to enable it or not)
- you say the backend logic is active so if option is enabled by customer, the order will be automatically split when confirmed?
- maybe we should also make sure this is correctly being displayed in back-office? merchant should see that the order has been split to allow available products being shipped first?

ping @PrestaShop/product-team who might want to discuss about the behavior :)

**Comment by LouiseBonnard** [comment]
@matks, @thalassa-web, following the current split order behavior in the BO, the two orders should share the name reference.

File: classes/checkout/CheckoutDeliveryStep.php
Comment (Progi1984) [inline classes/checkout/CheckoutDeliveryStep.php:40]:
```suggestion
    /**
     * @var bool
     */
    private $separatePackagesAllowed = false;
```

File: classes/checkout/CheckoutDeliveryStep.php
Comment (Progi1984) [inline classes/checkout/CheckoutDeliveryStep.php]:
```suggestion
    public function setSeparatePackagesAllowed(bool $separatePackagesAllowed): self
```

File: classes/checkout/CheckoutDeliveryStep.php
Comment (Progi1984) [inline classes/checkout/CheckoutDeliveryStep.php]:
```suggestion
    public function isSeparatePackagesAllowed(): bool
```

File: classes/checkout/CheckoutDeliveryStep.php
Comment (Progi1984) [reply classes/checkout/CheckoutDeliveryStep.php]:
```suggestion
                $requestParams['allow_separated_package']) ?? false
```

File: classes/checkout/CheckoutSession.php
Comment (Progi1984) [inline classes/checkout/CheckoutSession.php]:
```suggestion
    public function isSeparatePackagesAllowed(): bool
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacls

**Comment by Julievrz** [comment]
> This pull request seems to contain new translation strings. I have summarized them below to ease up review:
> 
> * `Shop.Theme.Checkout`
>   
>   * [ ]  [`Send available products first`](https://github.com/PrestaShop/PrestaShop/pull/22941/files#diff-ce8fe45c8932c4dfa982ef70fb842fbdR113)
> 
> (Note: this is an automated message, but answering it will reach a real human)

As the wording will appear in the front office, doesn't it make more sense to write "Receive available products first" in Shop.Theme.Checkout?

**Comment by Hlavtox** [comment]
@thalassa-web @matks This is an amazing feature! 🚀

**Comment by MatShir** [comment]
@thalassa-web very cool feature, I just add waiting for UX to be sure that the location of the feature matches well with our checkout, but you still can continue on your work 🚀

**Comment by prestascott** [comment]
Hello everyone, 

Thanks @thalassa-web for your PR! It's a great improvement. :)
Placement of the option to receive available products first is fine in the Shipping method step. I just recommend to put the option just before the comment section because it's more important, it needs to be placed higher in the information hierarchy. Comment section is used to allow customers to write additional information about the shipping, and split order can be a topic where he can write more about. 

Thanks!

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (Julievrz) [inline themes/classic/templates/checkout/_partials/steps/shipping.tpl]:
```suggestion
                <label for="allow_seperated_package">{l s='Receive available products first' d='Shop.Theme.Checkout'}</label>
```

**Comment by marionf** [comment]
Hello @thalassa-web 

It sounds very good to restore the checkbox "Send available products first" at the shipping step during the checkout process :+1: 
Could you handle [the feedback](https://github.com/PrestaShop/PrestaShop/pull/22941#issuecomment-774138671)  of @prestascott and the code & wording changes requested ?

**Comment by kpodemski** [comment]
Hello @thalassa-web 

Good work, do you plan to address feedback from our team? It would be great to have this feature back

**Comment by matks** [comment]
Ping @thalassa-web friendly reminder 😉

**Comment by thalassa-web** [comment]
Hi ! Sorry didn't see the ping before now … 
I'll try to do the update as soon as I can ;)

**Comment by thalassa-web** [comment]
Hi everyone …
Sorry to respond just now … did'nt receive any e-mail about comments on this PR … (I should check my email config …)
Just to be sure, because none of my previous PR was sooo followed and with some changes, I need some help to know what I have to do with this PR …
Specificly with my IDE and git …

**Comment by PierreRambaud** [comment]
> Hi everyone …
> Sorry to respond just now … did'nt receive any e-mail about comments on this PR … (I should check my email config …)
> Just to be sure, because none of my previous PR was sooo followed and with some changes, I need some help to know what I have to do with this PR …
> Specificly with my IDE and git …

You can peek proposed changes, add them into your branch, and push it :)

**Comment by thalassa-web** [comment]
Hi guys !
Ok it seems I found how to include suggestions in the PR (thanks to PierreRambaud).
One last question though, the bot suggested there is a new translation string (and indeed, there is).
Where can I put the new string «Receive available products first» for Shop.Theme.Checkout ?
(I could only translate it in French ^-^)

**Review by Julievrz — DISMISSED** [review]
Wording ✅

File: classes/checkout/CheckoutDeliveryStep.php
Comment (PierreRambaud) [inline classes/checkout/CheckoutDeliveryStep.php]:
```suggestion
                $requestParams['allow_separated_package'] ?? false
```

File: classes/checkout/CheckoutSession.php
Comment (PierreRambaud) [inline classes/checkout/CheckoutSession.php]:
```suggestion
    public function setSeparatePackagesAllowed(bool $separatePackagesAllowed): bool
```
?

File: classes/checkout/CheckoutSession.php
Comment (PierreRambaud) [inline classes/checkout/CheckoutSession.php]:
```suggestion
        $this->context->cart->allow_separated_package = $separatePackagesAllowed;
```

**Review by Julievrz — DISMISSED** [review]

**Comment by thalassa-web** [comment]
@matks waiting for some news about this PR …

**Comment by kpodemski** [comment]
Hello @thalassa-web 

Could you fix the issues coming from cs-fixer? :-)

File: classes/checkout/CheckoutDeliveryStep.php
Comment (kpodemski) [inline classes/checkout/CheckoutDeliveryStep.php]:
```suggestion
                    'isSeparatePackage' => $this->getCheckoutSession()->isSeparatePackagesAllowed(),
```

**Comment by thalassa-web** [comment]
It's quite annoying … I haven't made any changing about wording with the last commit, but I have to ask the approval again …

**Review by Julievrz — DISMISSED** [review]

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (kpodemski) [inline themes/classic/templates/checkout/_partials/steps/shipping.tpl]: @thalassa-web can we change this class? I think it should be related to the separated package checkbox

Comment (kpodemski) [inline themes/classic/templates/checkout/_partials/steps/shipping.tpl]:
```suggestion
              <input id="allow_separated_package" name="allow_separated_package"
```

In the end, I don't we need it

**Comment by Julievrz** [comment]
> It's quite annoying … I haven't made any changing about wording with the last commit, but I have to ask the approval again …

I know :-( The bot is not always right. (Maybe @atomiix has an idea?) I approved the wording of this PR again.

**Comment by kpodemski** [comment]
@Julievrz, I think it would be pretty complicated to handle this, but maybe we can agree that one of the maintainers can set this label to `wording ok` if nothing changes?

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (kpodemski) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl]: but don't trust me with this one, if there are some visual aspects related to this class, we need to make sure that we're add a dedicated one

Comment (thalassa-web) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl]: Honestly, I think it comes from the source code I copied from PrestaShop 1.6.
This PR has almost 1 year now … and some reviewed was made.
If the class is not used, I know it makes some dead code … but It worked as this … Now, I have to check if the removal of the class is ok or not. It will dellay (again) the resolution for just a class attribute …

Comment (thalassa-web) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl]: There is some js code which use this class : 
themes/classic/_dev/js/checkout.js
themes/classic/assets/js/theme.js
tests/Selenium/specs/checkout.js

So I don't think we have to remove the class.

Comment (thalassa-web) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl]: Or maybe I missused the class …
Is it the class for the gift checkbox maybe ?
Copy/Paste error ?

Comment (thalassa-web) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl]: I think you're right … We should remove this class …

Comment (thalassa-web) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl]: After some discussion with myself and I … You're right !

**Review by thalassa-web — COMMENTED** [review]
I do not know yet … What is the point with the Review …

**Comment by kpodemski** [comment]
@thalassa-web I know that it is a bit annoying that it all takes so much time. Thank you very much for your patience and contribution 💪

**Comment by kpodemski** [comment]
@thalassa-web, I don't think your change is enough to make this mechanism work after further checking this topic. There's much more to it. I'm afraid that it would be hard to make it work with 1.7 without additional work :-(

**Comment by thalassa-web** [comment]
@kpodemski I tried 1 year ago on a 1.7.x and it worked well because lot of work was already made since the 1.6 version …
I tried on my PerstaShop and the order was splitted …
Can you give me some clues about why it won't work ?

**Comment by kpodemski** [comment]
@thalassa-web I just double-checked `Cart` class, and there's indeed an `allow_seperated_package`, I must have searched for the wrong phrase. Sorry for the confusion

**Comment by atomiix** [comment]
@thalassa-web @Julievrz the label "Waiting for wording" is added each time a PR contains a wording (new or old) and each time a new commit is made as it can change the context of the wording and may require a new validation. The post made by prestonbot with checkboxes though, should keep track of validated new wording even when new commits are made. From what I see here, @Julievrz checked the checkbox **at the same time** prestonbot updated the comment, that's probably why it didn't stay checked.

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (PierreRambaud) [inline themes/classic/templates/checkout/_partials/steps/shipping.tpl]:
```suggestion
            {if $separatePackages.allowed}
              <span class="custom-checkbox">
                <input id="allow_separated_package" name="allow_separated_package"
                  type="checkbox" value="1" {if $separatePackages.isSeparatePackage}checked="checked"{/if}>
                <span><i class="material-icons rtl-no-flip checkbox-checked">&#xE5CA;</i></span>
                <label for="allow_separated_package">{l s='Receive available products first' d='Shop.Theme.Checkout'}</label>
              </span>
            {/if}
```

Comment (PierreRambaud) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl]: Indentation seems broken :(

**Review by PierreRambaud — DISMISSED** [review]

File: classes/checkout/CheckoutSession.php
Comment (Progi1984) [inline classes/checkout/CheckoutSession.php:209]:
```suggestion
        return $this->context->cart->allow_separated_package;
```

Comment (thalassa-web) [reply classes/checkout/CheckoutSession.php:209]: Why not but in this case we have to change the property in the Cart class …
Or I didn't change this !!!
If I commit this, the fonctionnality will break …

Comment (thalassa-web) [reply classes/checkout/CheckoutSession.php:209]: It's an existing typo error in the PrestaShop project …
Check on the main branch ;)

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (Progi1984) [inline themes/classic/templates/checkout/_partials/steps/shipping.tpl:85]:
```suggestion
                  type="checkbox" value="1"{if $separatePackages.isSeparatePackage} checked="checked"{/if} />
```

**Review by Progi1984 — COMMENTED** [review]
@thalassa-web Thank you for your contribution. I have detected a small bug. Could you fix it ?

File: classes/checkout/CheckoutSession.php
Comment (Progi1984) [reply classes/checkout/CheckoutSession.php:209]: Ok. In this case the l. 202 is not good ;)

**Comment by thalassa-web** [comment]
@Progi1984 Which one ? The one with the input without the closing tag ?

File: classes/checkout/CheckoutSession.php
Comment (Progi1984) [inline classes/checkout/CheckoutSession.php]: @thalassa-web This one is the bad property :

```suggestion
        $this->context->cart->allow_seperated_package = $separatePackagesAllowed;
```

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (thalassa-web) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl:85]: I am not against to fix errors, but I begin to think you do not want of this PR.
Indeed, I completed the tpl file with some copy/paste of the original code …
In the same file, we have inputs radio with this line of code :
```html
<input type="radio" name="delivery_option[{$id_address}]" id="delivery_option_{$carrier.id}"
                               value="{$carrier_id}"{if $delivery_option == $carrier_id} checked{/if}>
```
I think it's good to respect xhtml but it is also good to have consistency in rules …

File: classes/checkout/CheckoutSession.php
Comment (thalassa-web) [reply classes/checkout/CheckoutSession.php]: You are right … I missread this …
I thought I used the good property but I did not …
Thanks

File: themes/classic/templates/checkout/_partials/steps/shipping.tpl
Comment (Progi1984) [reply themes/classic/templates/checkout/_partials/steps/shipping.tpl:85]: It was a bonus not a Must Have ;)

**Review by Progi1984 — APPROVED** [review]

**Review by thalassa-web — COMMENTED** [review]
It seems review is ok

**Review by atomiix — APPROVED** [review]

**Comment by marwachelly** [comment]
Hello,
I tested the following case and I noticed that the feature of splitting order is not implemented:
> Go to FO choose a customized product
> Create an order with the same customized product with different customization and different quantity (set the quantity desired in the shopping cart)
Ps: the total sum of the quantity exceeds the stock
> Check the checkbox Receive available products first
> Go to BO > order page > The order is not splitted 
image

image

image

https://www.awesomescreenshot.com/video/5881863?key=b72664897be5f691044bb94f6dff9532

Could you please check.
Thanks,

**Comment by matks** [comment]
Hi,

Since we had no news from you for more than 30 days, I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale. Feel free to reopen or open another one if you think it's still relevant 😉 . You can ping me!

Thanks! and maybe see you soon

**Pull request closed without merging**
